### PR TITLE
Updates to include HTMHF triggers in GT emulator

### PR DESCRIPTION
**Pull request by elfontan**
#### PR description:
This PR includes updates to include HTMHF triggers in GT emulator for the version 2024_v1_3_0 of the L1 menu (end of 2024 pp data-taking), as detailed in the following:
* Inclusion of HTMHF object type in ESum in the GT data formats
* Updates of the TriggerMenuParser to properly parse the new condition from the xml menu (according to new utm libraries utm_0.13.0: see [code](https://gitlab.cern.ch/cms-l1t-utm/utm/-/tree/utm_0.13.0?ref_type=tags) and [documentation](https://globaltrigger.web.cern.ch/globaltrigger/release/utm/latest_doc/html/releaseNotes/v0_13_0.html)).
* Propagation of the HTMHF trigger implementation to EnergySum condition (and correlations)
* Propagation of the type definition changes to DataFormats/HLTReco and HLTL1TSeed module

#### PR validation:
The PR has been prepared starting from CMSSW_14_1_X_2024-06-23-0000:
```
cmsrel CMSSW_14_1_X_2024-06-23-0000
cd CMSSW_14_1_X_2024-06-23-0000/src
cmsenv
git cms-init
git cms-addpkg L1Trigger/L1TGlobal
git cms-addpkg DataFormats/L1TGlobal
git cms-addpkg DataFormats/HLTReco
git cms-addpkg HLTrigger/HLTfilters
scram b && scram b code-checks && scram b code-format && scram b
```

Test vectors have been produced by running the GT emulator standalone on a single orbit (3564 events) of EphemeralHLTPhysics 2024E (run 381065). Counts from the newly integrated triggers L1_MHTHF* in [L1Menu_Collisions2024_v1_3_0](https://github.com/cms-l1-dpg/L1MenuRun3/tree/master/development/L1Menu_Collisions2024_v1_3_0) are obseved:

| L1 seeds    | Counts |
| ---------- | ------- |
|L1_MHTHF100 | 179 |
|L1_MHTHF120 | 99 |
|L1_MHTHF125 | 86 |   
|L1_MHTHF130 | 72 |
|L1_MHTHF150 | 44 |

#### Backport:
Backport to CMSSW_14_0_X needed by the time of the deployment of the next L1+HLT menus: https://github.com/cms-sw/cmssw/pull/45356.

[cc: @eyigitba @slaurila @caruta @cavana @nabrandman ]

### CODE REVIEW

**Comment by epalencia** [comment]
Please test

File: DataFormats/L1TGlobal/interface/GlobalObject.h [hunk 1]
```diff
@@ -59,7 +59,8 @@ namespace l1t {
     gtExternal = 28,
     gtZDCP = 29,
     gtZDCM = 30,
-    ObjNull = 31,
+    gtHTMHF = 31,
+    ObjNull = 32,
```
Comment (missirol) [inline DataFormats/L1TGlobal/interface/GlobalObject.h:63]: To the best of my understanding, and as written at the top of this file, existing entries should not be modified, and new entries should be assigned new numbers.
```suggestion
    ObjNull = 31,
    gtHTMHF = 32,
```

File: DataFormats/L1TGlobal/interface/GlobalObject.h [hunk 2]
```diff
@@ -94,7 +95,8 @@ namespace l1t {
       {gtExternal, "External"},        // 28
       {gtZDCP, "ZDCP"},                // 29
       {gtZDCM, "ZDCM"},                // 30
-      {ObjNull, "ObjNull"},            // 31
+      {gtHTMHF, "HTMHF"},              // 31
+      {ObjNull, "ObjNull"},            // 32
```
Comment (missirol) [inline DataFormats/L1TGlobal/interface/GlobalObject.h:99]:
```suggestion
      {ObjNull, "ObjNull"},            // 31
      {gtHTMHF, "HTMHF"},              // 32
```

File: DataFormats/L1TGlobal/test/test_catch2_l1tGlobalObject.cc [hunk 3]
```diff
@@ -99,6 +99,7 @@ TEST_CASE("Test l1t::GlobalObject", "[l1tGlobalObject]") {
     REQUIRE(isValidGlobalObject(28, l1t::gtExternal, "External"));        // 28
     REQUIRE(isValidGlobalObject(29, l1t::gtZDCP, "ZDCP"));                // 29
     REQUIRE(isValidGlobalObject(30, l1t::gtZDCM, "ZDCM"));                // 30
-    REQUIRE(isValidGlobalObject(31, l1t::ObjNull, "ObjNull"));            // 31
+    REQUIRE(isValidGlobalObject(31, l1t::gtHTMHF, "HTMHF"));              // 31
+    REQUIRE(isValidGlobalObject(32, l1t::ObjNull, "ObjNull"));            // 32
```
Comment (missirol) [inline DataFormats/L1TGlobal/test/test_catch2_l1tGlobalObject.cc:103]:
```suggestion
    REQUIRE(isValidGlobalObject(31, l1t::ObjNull, "ObjNull"));            // 31
    REQUIRE(isValidGlobalObject(32, l1t::gtHTMHF, "HTMHF"));              // 32
```

**Review by missirol — COMMENTED** [review]
Looking at downstream classes, I would say that `HLTL1TSeed`  and [`TriggerTypeDefs`](https://github.com/missirol/cmssw/blob/265462ac63ac7b4730432ae33cb07708837d9d50/DataFormats/HLTReco/interface/TriggerTypeDefs.h) also need to be updated in order to support the new `HTMHF` objects. @cms-sw/hlt-l2

**Comment by mmusich** [comment]
assign hlt

File: DataFormats/L1TGlobal/interface/GlobalObject.h (hunk 1 above)
Comment (elfontan) [reply DataFormats/L1TGlobal/interface/GlobalObject.h:63]: Hi @missirol thank you! I misunderstood the description then, I thought taht ObjNull was not part of the problem and we simply needed to reassign a certain label to a different GT object. I don't fully understand when we should expect to have ObjNull (and why it should be a problem to assign that label to a new object, if the old one was never used...).
I can change this but I am wondering if it's not confusing to have ObjNull not as last one.

**Comment by elfontan** [comment]
> Looking at downstream classes, I would say that `HLTL1TSeed` and [`TriggerTypeDefs`](https://github.com/missirol/cmssw/blob/265462ac63ac7b4730432ae33cb07708837d9d50/DataFormats/HLTReco/interface/TriggerTypeDefs.h) also need to be updated in order to support the new `HTMHF` objects. @cms-sw/hlt-l2

Hi @missirol, for the moment we didn't update the HLT part because we don't use (and plan to use, for the moment) these new seeds at HLT (@eyigitba @slaurila @caruta @lathomas ). (Then, could this create a problem at HLT in any way? Do you have any suggestion to test the full chain without having a HLT path using those seeds/objects?) 
Anyway, I can work on propagating the change to those files.

**Comment by mmusich** [comment]
for my own education what does "HTMHF" mean?

**Comment by elfontan** [comment]
> for my own education what does "HTMHF" mean?

Missing HT (with HF included)

**Comment by lathomas** [comment]
HTMHF = MHTHF = missing ht (defined as the magnitude of the vector sum of the pt of all jets, including HF jets).

**Comment by mmusich** [comment]
> Then, could this create a problem at HLT in any way? Do you have any suggestion to test the full chain without having a HLT path using those seeds/objects?

I don't think it would create problems at HLT online, others should correct me (provided we don't mess up offline analysis with the indexing, see also https://github.com/cms-sw/cmssw/pull/42634#discussion_r1302636113) , though as discussed during the TSG meeting is unclear what this update ultimately leads to. If there is any prospect of using at HLT, I would be in favor of adding the whole chain here.

**Comment by eyigitba** [comment]
> If there is any prospect of using at HLT, I would be in favor of adding the whole chain here.

Does this have to be a part of this PR? 

I don't want to put this on @elfontan who's only implementing this change as the uGT emulator expert. Maybe @lathomas can comment on whether he can help with the rest of the uGT -> HLT interface implementations.

**Comment by mmusich** [comment]
> Does this have to be a part of this PR?

as far as I understand the missing HLT part is akin to https://github.com/cms-sw/cmssw/pull/42707 (@missirol might correct me). The changes seem simple enough to not warrant a different PR.

**Comment by mmusich** [comment]
@cmsbuild, please test

**Comment by elfontan** [comment]
Dear @missirol @mmusich @Martin-Grunewald,
I added a commit to modify the index in the GlobalObject enum and avoid to change the one of ObjNull.
I also added and used a new type _TriggerL1HTMHF_ to be coherent with the _TriggerL1ETMHF_ naming.

However, I noticed that TriggerTypeDefs already include a Type called _TriggerL1MissingHtHF_ that it's not used anywhere and it's probably related to what we need (see [here](https://github.com/search?q=repo%3Acms-sw%2Fcmssw%20TriggerL1MissingHtHF&type=code)). It was included for HI in 2018 (see [here](https://github.com/cms-sw/cmssw/commit/0c780edc6f02bd33712ca3c208f1e906f6485f49)). 

Any preference on the way to proceed? Do we want to use the already present type even if naming might be confusing? Or do we want to clean the code from the unused one? Or are we happy with the current PR?
Let me know your thoughts!

Cheers,
--Elisa

**Comment by lathomas** [comment]
Hello. 
Indeed `TriggerL1MissingHtHF`  is exactly this. I guess only `HLTL1TSeed.cc` needs to be updated then (?). 

To put some context regarding the possible use at HLT in the future: this algo provides an alternative way to ETMHF to measure estimate missing Et at L1. Since it relies on jets instead of trigger towers directly, it is expected to be more PU resilient and better calibrated. For now, we will monitor the performances in the menu. If it turns out to perform well, the long term plan would be to replace ETMHF seeds (mostly used to in pure MET/MHT paths where L1 is driving the turnon) by these.

**Comment by elfontan** [comment]
Attempt to test L1+HLT chain with the following command leads to errors:
```
hltGetConfiguration /dev/CMSSW_14_0_0/GRun/V143 --globaltag 140X_dataRun3_HLT_v3  --data --unprescale --output minimal --max-events -1 --eras Run3 --l1-emulator uGT --l1Xml L1Menu_Collisions2024_v1_3_0.xml --input /store/data/Run2024D/EphemeralHLTPhysics1/RAW/v1/000/380/346/00000/193c1072-a606-45b6-889e-ca64fbde6a2a.root  > hlt.py
```

Error below:
```
%MSG                                                                                                                                                                  
----- Begin Fatal Exception 26-Jun-2024 00:21:58 CEST-----------------------                                                                                          
An exception of category 'ProductNotFound' occurred while                                                                                                             
   [0] Processing  Event run: 380346 lumi: 175 event: 271989599 stream: 3                                                                                             
   [1] Running path 'HLTAnalyzerEndpath'                                                                                                                              
   [2] Prefetching for module L1TRawToDigi/'hltGtStage2Digis'                                                                                                         
   [3] Prefetching for module RawDataCollectorByLabel/'rawDataCollector'                                                                                              
   [4] Calling method for module L1TDigiToRaw/'packGtStage2'                                                                                                          
Exception Message:                                                                                                                                                    
Principal::getByToken: Found zero products matching all criteria                                                                                                      
Looking for type: BXVector<l1t::EtSum>                                                                                                                                
Looking for module label: simCaloStage2Digis                                                                                                                          
Looking for productInstanceName:                                                                                                                                                                                                                                                                                                      
   Additional Info:                                                                                                                                                   
      [a] If you wish to continue processing events after a ProductNotFound exception,                                                                                
add "TryToContinue = cms.untracked.vstring('ProductNotFound')" to the "options" PSet in the configuration.                                                                                                                                                                                                                           
----- End Fatal Exception -------------------------------------------------                                   
```

**Comment by Martin-Grunewald** [comment]
> Dear @missirol @mmusich @Martin-Grunewald, I added a commit to modify the index in the GlobalObject enum and avoid to change the one of ObjNull. I also added and used a new type _TriggerL1HTMHF_ to be coherent with the _TriggerL1ETMHF_ naming.
> 
> However, I noticed that TriggerTypeDefs already include a Type called _TriggerL1MissingHtHF_ that it's not used anywhere and it's probably related to what we need (see [here](https://github.com/search?q=repo%3Acms-sw%2Fcmssw%20TriggerL1MissingHtHF&type=code)). It was included for HI in 2018 (see [here](https://github.com/cms-sw/cmssw/commit/0c780edc6f02bd33712ca3c208f1e906f6485f49)).
> 
> Any preference on the way to proceed? Do we want to use the already present type even if naming might be confusing? Or do we want to clean the code from the unused one? Or are we happy with the current PR? Let me know your thoughts!
> 
> Cheers, --Elisa

Indeed, it should be safer to add HTMHF (as the other one was already foreseen [and used?] for a different purpose).

**Comment by Martin-Grunewald** [comment]
please test

**Comment by mmusich** [comment]
I guess there is no point in testing,  looks like https://github.com/cms-sw/cmssw/pull/45037 created massive failures in IBs

**Comment by missirol** [comment]
@elfontan 

One more comment. Maybe it's beyond the scope of this PR, but it seems to me that the new "HTMHF" object may need to be propagated downstream in a few places for DQM and NanoAOD.

Examples.
 - https://github.com/cms-sw/cmssw/blob/c0b15502e530da86f3fada015ba93ccd5ece93e3/DQMOffline/L1Trigger/src/L1TStage2CaloLayer2Offline.cc#L242-L262
 - https://github.com/cms-sw/cmssw/blob/c0b15502e530da86f3fada015ba93ccd5ece93e3/PhysicsTools/NanoAOD/plugins/TriggerObjectTableProducer.cc#L234-L237

Actually, in the NanoAOD part (e.g.
https://github.com/cms-sw/cmssw/blob/c0b15502e530da86f3fada015ba93ccd5ece93e3/PhysicsTools/NanoAOD/plugins/TriggerObjectTableProducer.cc#L204-L207
), I do not understand why `TriggerL1ETM` is being used for `"L1ETMHF"`. Naively it should be `TriggerL1ETMHF`, but maybe I'm missing something.

**Comment by mmusich** [comment]
@cmsbuild, please test

**Comment by mmusich** [comment]
> -1

still unrelated failures, will be fixed by https://github.com/cms-sw/cmssw/pull/45288

**Comment by mmusich** [comment]
@cmsbuild, please test

**Comment by elfontan** [comment]
> @elfontan
> 
> One more comment. Maybe it's beyond the scope of this PR, but it seems to me that the new "HTMHF" object may need to be propagated downstream in a few places for DQM and NanoAOD.
> 
> Examples.
> 
>     * https://github.com/cms-sw/cmssw/blob/c0b15502e530da86f3fada015ba93ccd5ece93e3/DQMOffline/L1Trigger/src/L1TStage2CaloLayer2Offline.cc#L242-L262
> 
>     * https://github.com/cms-sw/cmssw/blob/c0b15502e530da86f3fada015ba93ccd5ece93e3/PhysicsTools/NanoAOD/plugins/TriggerObjectTableProducer.cc#L234-L237
> 
> 
> Actually, in the NanoAOD part (e.g.
> 
> https://github.com/cms-sw/cmssw/blob/c0b15502e530da86f3fada015ba93ccd5ece93e3/PhysicsTools/NanoAOD/plugins/TriggerObjectTableProducer.cc#L204-L207
> 
> ), I do not understand why `TriggerL1ETM` is being used for `"L1ETMHF"`. Naively it should be `TriggerL1ETMHF`, but maybe I'm missing something.

Hi @missirol thank you for your point!
I personally believe that this can be kept in mind by the L1 group (@epalencia @aloeliger @eyigitba @slaurila) as later update and is not essential to move on with this update, for now. (We could maybe open a github issue to keep track?) 
But I let them comment more on this.

I wait for further feedback to proceed with a backport of the changes to CMSSW_14_0_X.

Cheers,
--Elisa

**Comment by slaurila** [comment]
> > @elfontan
> > One more comment. Maybe it's beyond the scope of this PR, but it seems to me that the new "HTMHF" object may need to be propagated downstream in a few places for DQM and NanoAOD.
> > Examples.
> > ```
> > * https://github.com/cms-sw/cmssw/blob/c0b15502e530da86f3fada015ba93ccd5ece93e3/DQMOffline/L1Trigger/src/L1TStage2CaloLayer2Offline.cc#L242-L262
> > 
> > * https://github.com/cms-sw/cmssw/blob/c0b15502e530da86f3fada015ba93ccd5ece93e3/PhysicsTools/NanoAOD/plugins/TriggerObjectTableProducer.cc#L234-L237
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Actually, in the NanoAOD part (e.g.
> > https://github.com/cms-sw/cmssw/blob/c0b15502e530da86f3fada015ba93ccd5ece93e3/PhysicsTools/NanoAOD/plugins/TriggerObjectTableProducer.cc#L204-L207
> > 
> > ), I do not understand why `TriggerL1ETM` is being used for `"L1ETMHF"`. Naively it should be `TriggerL1ETMHF`, but maybe I'm missing something.
> 
> Hi @missirol thank you for your point! I personally believe that this can be kept in mind by the L1 group (@epalencia @aloeliger @eyigitba @slaurila) as later update and is not essential to move on with this update, for now. (We could maybe open a github issue to keep track?) But I let them comment more on this.
> 
> I wait for further feedback to proceed with a backport of the changes to CMSSW_14_0_X.
> 
> Cheers, --Elisa

Hi, in the interest of time, I'd suggest we proceed with this PR and the backport ASAP (to enable the HLT menu development), and save the DQM and Nano updates for separate PRs that can follow later. @epalencia @aloeliger can you please comment if you also think this is fine, i.e. nothing breaks in L1 Nano or DQM even if we don't update them right now.

Thanks!
Santeri

**Comment by aloeliger** [comment]
> > > @elfontan
> > > One more comment. Maybe it's beyond the scope of this PR, but it seems to me that the new "HTMHF" object may need to be propagated downstream in a few places for DQM and NanoAOD.
> > > Examples.
> > > ```
> > > * https://github.com/cms-sw/cmssw/blob/c0b15502e530da86f3fada015ba93ccd5ece93e3/DQMOffline/L1Trigger/src/L1TStage2CaloLayer2Offline.cc#L242-L262
> > > 
> > > * https://github.com/cms-sw/cmssw/blob/c0b15502e530da86f3fada015ba93ccd5ece93e3/PhysicsTools/NanoAOD/plugins/TriggerObjectTableProducer.cc#L234-L237
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Actually, in the NanoAOD part (e.g.
> > > https://github.com/cms-sw/cmssw/blob/c0b15502e530da86f3fada015ba93ccd5ece93e3/PhysicsTools/NanoAOD/plugins/TriggerObjectTableProducer.cc#L204-L207
> > > 
> > > ), I do not understand why `TriggerL1ETM` is being used for `"L1ETMHF"`. Naively it should be `TriggerL1ETMHF`, but maybe I'm missing something.
> > 
> > 
> > Hi @missirol thank you for your point! I personally believe that this can be kept in mind by the L1 group (@epalencia @aloeliger @eyigitba @slaurila) as later update and is not essential to move on with this update, for now. (We could maybe open a github issue to keep track?) But I let them comment more on this.
> > I wait for further feedback to proceed with a backport of the changes to CMSSW_14_0_X.
> > Cheers, --Elisa
> 
> Hi, in the interest of time, I'd suggest we proceed with this PR and the backport ASAP (to enable the HLT menu development), and save the DQM and Nano updates for separate PRs that can follow later. @epalencia @aloeliger can you please comment if you also think this is fine, i.e. nothing breaks in L1 Nano or DQM even if we don't update them right now.
> 
> Thanks! Santeri

@slaurila in the interest of making the HLT deadline, I think this is fine with me

**Comment by aloeliger** [comment]
+l1

**Comment by Martin-Grunewald** [comment]
@elfontan 
Please prepare the PR to backport this to 14_0. Thanks!

**Comment by Martin-Grunewald** [comment]
+1

**Comment by mmusich** [comment]
@elfontan @slaurila @eyigitba please open a backport ASAP.

**Comment by rappoccio** [comment]
+1

**Pull request merged**